### ISSUE TITLE
Ipad/Iphone devices InitialSpeechInstruction not working as expected.

### ISSUE THREAD

**Issue by karanakatle** [aws-samples/aws-lex-web-ui#404]
Hi Team,

We are facing this issue only on Ipad/Iphone devices not on android and desktop. So when we open a chatbot and click on the mic, the initialspeechinstruction is readed. As soon as bot goes to listening mode, it again reads the InitialSpeechInstructions. Not able to understand why this is happening in Ipad/Iphone only and not in any other device.

Any comments on this?

**Comment by chrislott** [comment]
Hi @karanakatle,

After looking at the code, this will only play if the user is logged in. On the iOS devices, is the user logged in before clicking microphone?  Is it different on the other devices?

**Comment by bobpskier** [comment]
@chrislott @karanakatle I was able to reproduce as a logged in user on iOS devices. I also reproduced on the latest version of Safari on OSX. It seems to  be more of a Safari/WebKit change than anything else. I also tried chrome on the same iOS device and it has the same problem. Chrome uses Apple's WebKit under the covers so I'm not surprised. Firefox on OSX does not exhibit the same problem where Safari does. 

The algorithm is that on the first click on the microphone icon, the lex-web-ui will initialize the audio and play a silent sound. This forces the popup to be shown asking the user if they will grant access to the microphone. After clicking yes, the initial speech will be played and the ui then waits for the user's voice input. 

It seems some change in Apple's WebKit is allowing the UI to set the startup speech input to be used next prior to complete playing the silent sound. 

I'll take a look to see what I can figure out.

**Comment by karanakatle** [comment]
Thank you @chrislott  @bobpskier , we are also trying to find a fix/workaround for this. If found, will love to share our thoughts here and get your comments on it.
Also @chrislott , we were thinking to remove isloggedin part from the logic in order to make it available for non logged in users also as user need not to be logged in for using our chatbot.

**Comment by bobpskier** [comment]
@karanakatle Allowing use of the cognito unauthenticated role to generate a polly response for the initial speech opens up the use of polly by anyone who understands how to use cognito within your AWS account. You really want to avoid opening this up from a security perspective. One solution that will take a bit of work is to precreate the polly audio manually and embed this blob into a custom lex-web-ui resource. Instead of calling polly to generate the audio the precreated audio could be embedded instead. This is probably how best to accomplish this.

**Comment by karanakatle** [comment]
@bobpskier, thnx for the suggestion and I guess this will solve our issues on iphone/ipad as well. In a code can you please provide a reference of how to create it manually. I am little bit new to JS, so if got a reference, it will be of great help.

**Comment by karanakatle** [comment]
@bobpskier thank you so much

**Comment by bobpskier** [comment]
@karanakatle I am able to automate creation of the initial speech during the codebuild through cloudformation into an mp3, place this in the s3 webapp bucket, and then have this played when voice interaction first starts up. This removes the requirement to have polly available in the unauth role. I need to clean up the implementation a bit and will then submit a pull request hopefully tomorrow. 

This does not solve the Safari issue first seen in version 15.5 of Safari. Safari/webkit appears to play the configured audio element twice. Once prior to asking for authorization to use then microphone and then again after the user clicks the allow button. I've not found a way to work around this problem. I'm still trying to debug a bit in Safari but at this point I'm not hopeful there is an easy solution.

**Comment by karanakatle** [comment]
@bobpskier thanks for the update. Is it possible to move the allow mic prompt in the first click of mic itself? Meaning as soon as we click on mic to generate initial speech instruction, allow mic prompt appears and further after clicking okay bot says the initial utterance.
Can this be a possible solution? Any comments?

**Comment by bobpskier** [comment]
@karanakatle Its probably not a good order to click on the mic and have the code start listening while then playing back the audio through the speakers. Too much interference. The code specifically initiates playback of the audio and then starts up the conversation through the mic. On my system, the first playback completes prior to the microphone approval request popping up. Not sure why Safari plays back the audio a second time after it has successfully played it the first time. Its a change in Safari behavior as prior versions work correctly. All other browsers I've tested work properly.

**Comment by bobpskier** [comment]
@karanakatle I was able to get the behavior back to normal on Safari by introducing a 100 msec delay between playback of the speech and start of the conversation. I need to perform some additional testing over multiple browsers and make documentation updates but I think the changes I have will work to a) precreate an mp3 for initial speech which allows the unauth role to playback initial speech without polly access and b) removes the repeat playback in Safari of the initial speech.

**Comment by karanakatle** [comment]
@bobpskier ohh that's great news, thanks for the update.
We also tried a couple of ways but was only able to handle a) scenario.
We will wait for your pull request. Once it is merged, will take the latest code.

**Comment by karanakatle** [comment]
@bobpskier is there any ETA for this? It will be helpful for us in deciding the next steps for testing and promoting to our envs.

**Comment by bobpskier** [comment]
@karanakatle I created a pull request with the changes. The AWS team will need to merge this into development branch. We'll need to test these changes before merging into master and releasing a new version of lex-web-ui. Is there any chance you can test using the changes in the pull request and let us know your results?

**Comment by karanakatle** [comment]
@bobpskier thank you so much, solution did worked for us. We used the delay part and audio.pause() (without audio.pause() it wasn't working) part in our solution and the issue of the initial utterance in iphone/ipad devices is resolved. 

However we are observing another issue, after the initial utterance if we give any voice input to the lex, the volume of the bot suddenly reduces while reading out next question. In iphone the volume gets too low to hear, in ipad we can observe the volume difference but it is not that much when compared with iphone. In android/desktop browser it is working fine. Only problem is present in iphone/ipad devices.

Is there any property you think in lex-web-ui which we can use to fix this stuff? Any comments on this?

**Comment by bobpskier** [comment]
@karanakatle I noticed the difference as well. Volume is under control of the user and device. I tried adjusting the volume property available on the audio element and this adjustment had no effect on the initial nor subsequent audio playback. No volume is set as far as I can see in the code and adding adjustments to volume did not have an effect on my devices. Both problems are due to Apple's implementation. It should not be necessary to use the audio.pause() but this tweak was the only way it seems to work around the first problem. We'd need to refer the issue to Apple and see what they say. Not sure I have any ins at Apple. I can try and submit a case on the apple developer site but I don't hold out much hope they'll respond. If you figure something out with volume please keep me in the loop. Thanks.
